_drafts/assert.md

```diff
@@ -0,0 +1,605 @@
+    if (!(expr)) {                        \
+      my_assert(__FILENAME__, __LINE__);  \
+    }                                     \
+  } while (0)
+```
+
+With this change, we reduce our file string from
+`/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert3.c`
+to `assert3.c`, resulting in 68 bytes of code space savings per file which
+contains an assert.
+
+```
+$ arm-none-eabi-objdump -s -j '.text' build/nrf52.elf
+ ...
+26b0 61737365 7274342e 6300               assert4.c
+```
+
+#### Extracted Information
+
+We know the file and line number of the assert hit.
+
+### Register Values Only Assert (#5)
+
+The problem with all of the above assert implementations is that our code size
+is bloated due to storing string filenames, and also that our code size can
+change depending on how we name the files in our project. This encourages poor
+naming practices, as well as produces unexpected build sizes.
+
+My favorite way to implement asserts to record both the Program Counter and the
+Link Register within the macro. This provides the following benefits:
+
+- Each assert consume a fixed number of bytes, no matter the filepath, filename,
+  or number of asserts per file.
+- By capturing both the PC and LR, we can get an extra frame in our 'backtrace'.
+- If we really wanted to micro-optimize, we can choose not to record the LR and
+  save a couple more bytes.
+
+Below we define two macros, `GET_LR()` and `GET_PC()`, to retrieve the registers
+`pc` and `lr` respectively.
+
+```c
+#define GET_LR() __builtin_return_address(0)
+#define GET_PC(_a) __asm volatile ("mov %0, pc" : "=r" (_a))
+
+#define MY_ASSERT_RECORD()     \
+  do {                         \
+    void *pc;                  \
+    GET_PC(pc);                \
+    const void *lr = GET_LR(); \
+    my_assert(pc, lr);         \
+  } while (0)
+
+#define MY_ASSERT(exp)         \
+  do {                         \
+    if (!(exp)) {              \
+      MY_ASSERT_RECORD();      \
+    }                          \
+  } while (0)
+```
+
+In our `my_assert` implmementation, we then record the two register values in
+our struct. Since we aren't storing the file or line number information directly
+in our firmware, we can also save RAM and flash space. This could allow us to
+store many more asserts in our logs!
+
+```c
+void my_assert(uint32_t *pc, uint32_t *lr) {
+  // File and line deliberately left empty
+  g_assert_info.pc = (uint32_t)pc;
+  g_assert_info.lr = (uint32_t)lr;
+  __asm("bkpt 5");
+}
+```
+
+The drawback to this approach is that we'll have to use `addr2line` or a similar
+tool to go from our register values to the file and line information, but this
+can be added to a simple script and used by the entire team with ease.
+
+```
+$ arm-elf-linux-addr2line --exe build/nrf52.elf 0x2250
+/Users/tyler/dev/memfault/interrupt/example/debugging-asserts/impls/assert5.c:13
+```
+
+> NOTE: GCC and various optimization levels may choose to combine multiple
```

Comment:
maybe: Compiler optimizations such as function inlining may result in _different_ asserts being merged together.

Aside: Though I'm now wondering if this winds up being try if the inline asm like we have above for the PC winds up being inserted. Not sure the compiler will merge inline asm calls

Reply:
Fixed